crypto/provider_core.c

```diff
@@ -556,9 +563,25 @@ OSSL_PROVIDER *ossl_provider_new(OSSL_LIB_CTX *libctx, const char *name,
         template.init = init_function;
     }
 
+    if (params != NULL) {
+        int i;
+
+        for (i = 0; params[i].key != NULL; i++) {
+            if (params[i].data_type != OSSL_PARAM_UTF8_STRING)
+                continue;
+            if (ossl_provider_info_add_parameter(&template, params[i].key,
+                                                 (char *)params[i].data) <= 0)
+                return NULL;
+        }
+    }
+
     /* provider_new() generates an error, so no need here */
-    if ((prov = provider_new(name, template.init, template.parameters)) == NULL)
+    if ((prov = provider_new(name, template.init, template.parameters)) == NULL) {
+        sk_INFOPAIR_pop_free(template.parameters, infopair_free);
         return NULL;
+    }
+    /* Parameters are copied, let's free them */
+    sk_INFOPAIR_pop_free(template.parameters, infopair_free);
```

Comment:
Doesn't this lead to a double free in the case where parameters are supplied from a config file? In that case we are doing a shallow copy of the `INFOPAIR` stack from the stack held in `store->provinfo`. Subsequently (when the store gets freed), `ossl_provider_store_free` gets called which also frees the `INFOPAIR` stack.

Reply:
Again, I think no - as template is a local variable.

Reply:
But line 549 above pre-populates it with data from `store->provinfo`

Reply:
Fixed, thanks!